src/sql/reuse.rs

```diff
@@ -320,4 +362,173 @@ mod tests {
         assert!(!predicate_implies(&pa, &pc));
         assert!(predicate_implies(&pc, &pa));
     }
+
+    #[test]
+    fn complex_predicate_implication_or() {
+        use nom_sql::ConditionExpression::*;
+        use nom_sql::ConditionBase::*;
+        use nom_sql::Literal;
+
+        let pa = ComparisonOp(ConditionTree {
+            operator: Operator::Less,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(20.into())))),
+        });
+        let pb = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(60.into())))),
+        });
+        let pc = ComparisonOp(ConditionTree {
+            operator: Operator::Less,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(10.into())))),
+        });
+        let pd = ComparisonOp(ConditionTree {
+            operator: Operator::Greater,
+            left: Box::new(Base(Field(Column::from("a")))),
+            right: Box::new(Base(Literal(Literal::Integer(80.into())))),
+        });
+
+        // a < 20 or a > 80
+        let cp1 = LogicalOp(ConditionTree {
+            left: Box::new(pa.clone()),
+            right: Box::new(pb.clone()),
+            operator: Operator::Or
+        });
+
+        // a < 10 or a > 60
+        let cp2 = LogicalOp(ConditionTree {
+            left: Box::new(pc),
+            right: Box::new(pd),
+            operator: Operator::Or
+        });
+
+        // a < 80 or a > 20
+        let cp3 = LogicalOp(ConditionTree {
+            left: Box::new(pb),
+            right: Box::new(pa),
+            operator: Operator::Or
+        });
```

Comment:
Yeah, I realized that was the goal! The predicate is still always true though, since [-infinity, 80] and [20, infinity] overlap ;-) The `cp1` and `cp2` predicates are the ones that appear in the Finkelstein paper, if I read it correctly -- `cp3` doesn't. But I think it's fine -- the point here is not whether the predicate is always true, but whether the implications are correct (which they are).